### PR TITLE
EN6-77 CMS: Create Digital Assets API Endpoints

### PR DESCRIPTION
**Pull request by ffleandro**

### CODE REVIEW

**Comment by entando-service** [comment]
All is Well

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
All is Well

**Comment by entando-service** [comment]
All is Well

File: plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/model/ImageResource.java [hunk 1]
```diff
@@ -131,6 +137,35 @@ public void saveResourceInstances(ResourceDataBean bean) throws ApsSystemExcepti
...
+                    if (!ignoreKeysList.contains(tag.getTagName())) {
+                        logger.debug("Add Metadata with key: {}", tag.getTagName());
+                        meta.put(tag.getTagName(), tag.getDescription());
+                    } else {
+                        logger.debug("Skip Metadata key {}", tag.getTagName());
+                    }
+                }
+            }
+        } catch (ImageProcessingException ex) {
+            logger.error("Error reading metadata");
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/model/ImageResource.java]: These exceptions should be logged

File: plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/model/ImageResource.java [hunk 2]
```diff
@@ -131,6 +137,35 @@ public void saveResourceInstances(ResourceDataBean bean) throws ApsSystemExcepti
...
+        } catch (IOException ioex) {
+            logger.error("Error reading file");
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/model/ImageResource.java]: Log exception

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            resourceFile.setDescr(file.getOriginalFilename());
+            resourceFile.setMainGroup(group);
+            resourceFile.setResourceType(resourceType);
+            resourceFile.setCategories(categories.stream()
+                .map(code -> categoryManager.getCategory(code)).collect(Collectors.toList()));
+
+            ResourceInterface resource = resourceManager.addResource(resourceFile);
+            return convertResourceToDto(resourceType, resourceManager.loadResource(resource.getId()));
+        } catch (ApsSystemException e) {
+            throw new RestServerError("Error saving new resource", e);
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Do these exceptions get logged elsewhere? And confirm that this is the right type for the error

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                resourceFile.setCategories(categories.stream()
+                        .map(code -> categoryManager.getCategory(code)).collect(Collectors.toList()));
+            } else {
+                resourceFile.setCategories(resource.getCategories());
+            }
+
+            resourceManager.updateResource(resourceFile);
+            return convertResourceToDto(resource.getType(), resourceManager.loadResource(resourceId));
+        } catch (ApsSystemException e) {
+            throw new RestServerError("Error updating image resource", e);
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Should this get logged here?

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+        List<String> allowedExtensions;
+        if ("Image".equals(resourceType)){
+            allowedExtensions = imageAllowedExtensions;
+        } else if ("Attach".equals(resourceType)){
+            allowedExtensions = fileAllowedExtensions;
+        } else throw new RestServerError(String.format("Invalid resource type %s", resourceType), null);
+
+        if (!allowedExtensions.contains(mimeType)) {
+            throw new RestServerError(String.format("Invalid mime type: %s", mimeType), null);
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: This should get logged. At a minimum the fact that an invalid type was encountered.

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        mimeType = Optional.ofNullable(mimeType)
+                .map(t -> t.split("/")[1])
+                .orElseThrow(() -> new RestServerError("Invalid mime type", null));
+
+        List<String> allowedExtensions;
+        if ("Image".equals(resourceType)){
+            allowedExtensions = imageAllowedExtensions;
+        } else if ("Attach".equals(resourceType)){
+            allowedExtensions = fileAllowedExtensions;
+        } else throw new RestServerError(String.format("Invalid resource type %s", resourceType), null);
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: This probably shouldn't be inlined like this. Add brackets (we need to create a style guide that we can install in IDEs

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        return dimensions;
+    }
+
+    private FieldSearchFilter[] createSearchFilters(String type, RestListRequest requestList) {
+        List<FieldSearchFilter> filters = new ArrayList<>();
+        filters.add(
+            new FieldSearchFilter(IResourceManager.RESOURCE_TYPE_FILTER_KEY, type, false)
+        );
+
+        //TODO better way to convert attributes to ResourceManager properties?
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Is this TODO still an issue? Lets resolve or remove comment

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+                case "createdAt":
+                    attr = IResourceManager.RESOURCE_CREATION_DATE_FILTER_KEY;
+                    break;
+                case "updatedAt":
+                    attr = IResourceManager.RESOURCE_MODIFY_DATE_FILTER_KEY;
+                    break;
+                case "group":
+                    attr = IResourceManager.RESOURCE_MAIN_GROUP_FILTER_KEY;
+                    break;
+                default:
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Is this an error condition? Should get logged at a minimum i think

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+        return filters.stream().toArray(FieldSearchFilter[]::new);
+    }
+
+    private EntitySearchFilter createOrderFilter(RestListRequest requestList) {
+        String groupBy = requestList.getSort();
+
+        String key = IResourceManager.RESOURCE_DESCR_FILTER_KEY; //default
+
+        //TODO better way to convert attributes to ResourceManager properties?
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Resolve or remove comment

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+        }
+
+        EntitySearchFilter filter = new EntitySearchFilter(key, false);
+        filter.setOrder(requestList.getDirection().equals("desc") ?
+                EntitySearchFilter.DESC_ORDER : EntitySearchFilter.ASC_ORDER);
+
+        return filter;
+    }
+
+    //TODO support array of categories
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Resolve or remove

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        if ("Image".equals(resourceType)) {
+            return convertImageResourceToDto((ImageResource) resource);
+        } else if ("Attach".equals(resourceType)) {
+            return convertFileResourceToDto((AttachResource) resource);
+        } else {
+            throw new RestServerError(String.format("Invalid resource type: %s", resourceType), null);
+        }
+    }
+
+    //TODO better encapsulate dto conversion, perhaps create separate builder?
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Resolve or remove comment

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/web/resource/ResourcesController.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @ApiOperation(value = "LIST Resources", nickname = "listResources", tags = {"resources-controller"})
+    @ApiResponses(value = {
+            @ApiResponse(code = 201, message = "Created"),
+            @ApiResponse(code = 401, message = "Unauthorized")})
+    @GetMapping("/plugins/cms/assets")
+    @RestAccessControl(permission = Permission.CONTENT_EDITOR)
+    public ResponseEntity<PagedRestResponse<AssetDto>> listAssets(@RequestParam("type") String type, RestListRequest requestList) {
+        logger.debug("REST request - list image resources");
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/web/resource/ResourcesController.java]: Log parameters as well

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/web/resource/ResourcesController.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+    @ApiOperation(value = "CREATE Resource", nickname = "createResource", tags = {"resources-controller"})
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "OK"),
+            @ApiResponse(code = 401, message = "Unauthorized")})
+    @PostMapping("/plugins/cms/assets")
+    @RestAccessControl(permission = Permission.CONTENT_EDITOR)
+    public ResponseEntity<SimpleRestResponse<AssetDto>> createAsset(@RequestParam("type") String type, @RequestPart("file") MultipartFile file,
+                @RequestParam String group, @RequestParam String categories) {
+        logger.debug("REST request - create new resource");
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/web/resource/ResourcesController.java]: Log params

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+                    .size(instance.getFileLength())
+                    .dimensions(String.format("%dx%d px", dimensions.getDimx(), dimensions.getDimy()))
+                    .build());
+        }
+
+        return builder.build();
+    }
+
+    private FileAssetDto convertFileResourceToDto(AttachResource resource){
+        FileAssetDto.FileAssetDtoBuilder builder = FileAssetDto.builder()
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Very nice!

File: plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/model/ResourceInterface.java [hunk 15]
```diff
@@ -217,6 +217,15 @@
...
     public void saveResourceInstances(ResourceDataBean bean) throws ApsSystemException;
 
+    /**
+     * Ricava ed salva tutte le istanze associate ad una risorsa, valorizzando
+     * quest'ultima con i dati delle istanze ricavate.
+     *
+     * @param bean L'oggetto detentore dei dati della risorsa da inserire.
+     * @throws ApsSystemException In caso di eccezioni.
+     */
+    public void saveResourceInstances(ResourceDataBean bean, List<String> ignoreMetadataKeys) throws ApsSystemException;
```
Comment (joewhite101) [inline plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/model/ResourceInterface.java]: Give that the method is overloaded lets update the comment to English and note the difference between the two.

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java (hunk 3 above)
Comment (ffleandro) [reply plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: These exceptions are being logged in the `RestExceptionHandler`, however are being treated as `GENERIC_ERROR`.

Ideally there should be a generic exception where you can set an HTTP status code and a message. Today, if you don't create a specific handler for each Exception, you always get a `500/GENERIC_ERROR`

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java (hunk 4 above)
Comment (ffleandro) [reply plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: It is an unresolveable internal exception that shouldn't occur.
So throwing it wrapped in a known type and letting the generic exception handler log it, is the appropriate way to handle these kind of errors, in my opinion.

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java (hunk 11 above)
Comment (ffleandro) [reply plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Commented for future work?

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java (hunk 3 above)
Comment (ffleandro) [reply plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: They are being logged inside the `ResourceManager` before throwing the `ApsSystemException`.

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java (hunk 4 above)
Comment (ffleandro) [reply plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: They are being logged inside the `ResourceManager` before throwing the `ApsSystemException`.

File: plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java (hunk 14 above)
Comment (ffleandro) [reply plugins/entando-plugin-jacms/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/resource/ResourcesService.java]: Lombok FTW!
Can still be better with streams, pushing changes now.

**Comment by entando-service** [comment]
All is Well

**Comment by entando-service** [comment]
All is Well

**Pull request merged**